### PR TITLE
ISSUE-177 Parallelize SAM algorithm in mineraly.py using Joblib

### PR DESCRIPTION
**Pull request by aheermann**
Adds joblib functionality to mineral.py as specified in #177. Will be one possible parallelization option that can be specified in config file

### CODE REVIEW

**Comment by lewismc** [comment]
@aheermann thanks for opening this PR. Please resolve above conflicts. Thank you.

File: pycoal/mineral.py [hunk 1]
```diff
@@ -24,6 +24,42 @@
 import time
 import fnmatch
 import shutil
+from pycoal import mineral
+import multiprocessing
+from joblib import Parallel, delayed
```
Comment (lewismc) [inline pycoal/mineral.py:29]: joblib needs to be added to [setup.py](https://github.com/capstone-coal/pycoal/blob/master/setup.py#L38)

File: pycoal/mineral.py [hunk 2]
```diff
@@ -24,6 +24,42 @@
...
+
+
+def CalculatePixel(pixel, x, y, classified, library, threshold, resample, scores_file_name):
```
Comment (lewismc) [inline pycoal/mineral.py]: Provide documentation for the function. An example can be found [here](https://github.com/capstone-coal/pycoal/blob/master/pycoal/mineral.py#L39-L66).

Comment (lewismc) [reply pycoal/mineral.py]: Also, let think about whether `CalculatePixel` is a suitable name for the function. It appears to me it is _**calculating a confidence value of a classifed pixel**_?

File: pycoal/mineral.py [hunk 3]
```diff
@@ -24,6 +24,42 @@
...
+        # calculate spectral angles
+        # angles = SpectralAngleReplacement(resampled_pixel[numpy.newaxis, numpy.newaxis, ...], library.spectra)
+        angles = spectral.spectral_angles(resampled_pixel[numpy.newaxis, numpy.newaxis, ...], library.spectra)
+        # normalize confidence values from [pi,0] to [0,1]
+        for z in range(angles.shape[2]):
+            angles[0,0,z] = 1-angles[0,0,z]/math.pi
+        # get index of class with largest confidence value
+        index_of_max = numpy.argmax(angles)
+
+        # get confidence value of the classied pixel
```
Comment (lewismc) [inline pycoal/mineral.py]: Spelling confidence value of the _**classied**_ pixel

File: pycoal/mineral.py [hunk 4]
```diff
@@ -24,6 +24,42 @@
...
+
+
+    # if it is not a no data pixel
+    if not numpy.isclose(pixel[0], -0.005) and not pixel[0]==-50:
+
+        # resample the pixel ignoring NaNs from target bands that don't overlap
+        # TODO fix spectral library so that bands are in order
+        resampled_pixel = numpy.nan_to_num(resample(pixel))
+        # calculate spectral angles
+        # angles = SpectralAngleReplacement(resampled_pixel[numpy.newaxis, numpy.newaxis, ...], library.spectra)
```
Comment (lewismc) [inline pycoal/mineral.py]: If not in use please remove.

File: pycoal/mineral.py [hunk 5]
```diff
@@ -85,49 +121,31 @@ def SAM(image_file_name, classified_file_name, library_file_name, scores_file_na
     if scores_file_name is not None:
         # allocate a zero-initialized MxN array for the scores image
         scored = numpy.zeros(shape=(M,N), dtype=numpy.float64)
+        scoredJobLib = numpy.zeros(shape=(M,N), dtype=numpy.float64)
```
Comment (lewismc) [inline pycoal/mineral.py]: Please follow variable naming conventions defined in PEP8 e.g. `scored_job_lib`

File: pycoal/mineral.py [hunk 6]
```diff
@@ -85,49 +121,31 @@ def SAM(image_file_name, classified_file_name, library_file_name, scores_file_na
...
+    end = time.time()
+    seconds = end - start
+   
 
-                    if scores_file_name is not None:
-                        # store score value
-                        scored[x,y] = score
 
-    # save the classified image to a file
+    print("Time to run: {} seconds".format(seconds))
```
Comment (lewismc) [inline pycoal/mineral.py]: Please redirect to logging. An example can be seen [here](https://github.com/capstone-coal/pycoal/blob/master/pycoal/mineral.py#L87).

**Review by lewismc — CHANGES_REQUESTED** [review]
Hi @aheermann thanks for posting your PR. Please consider and address the comments in my review.

File: pycoal/mineral.py [hunk 7]
```diff
@@ -85,49 +121,31 @@ def SAM(image_file_name, classified_file_name, library_file_name, scores_file_na
...
+                        x, y, classified, library, threshold, resample, scores_file_name)
+                        for x in range(M) for y in range(N)), dtype=numpy.float64)
 
-                    # index from one (after zero for no data)
-                    classified[x,y] = index_of_max + 1
+    #puts it all in one single array, need to make 2d array
+    k = 0
+    for i in range(M):
+        for j in range(N):
+            if scoredSingle[k] is scoredSingle[k]:
```
Comment (lewismc) [inline pycoal/mineral.py]: Please follow variable naming conventions defined in PEP8 e.g. `scored_single`

File: pycoal/mineral.py [hunk 8]
```diff
@@ -85,49 +121,31 @@ def SAM(image_file_name, classified_file_name, library_file_name, scores_file_na
...
+                scoredJobLib[i][j] = scoredSingle[k]
```
Comment (lewismc) [inline pycoal/mineral.py]: same here

**Comment by lewismc** [comment]
@aheermann this is much better. Thanks for updating.
Please see above, specifically resolve conflict(s) in `pycoal/mineral.py`. 
Additionally, now that #161 is addressed via #178 you should merge your branch with master. All of the CI check above need to pass before we merge this into master.
Please let me know when I can test this out... right now I can't.
Thank you

**Comment by lewismc** [comment]
Excellent @aheermann this looks much better.
Can you please scope out the [test regressions](https://travis-ci.org/capstone-coal/pycoal/builds/605232345?utm_source=github_status).
We can discuss this on Friday if you want... I will investigate in the meantime as well.

**Review by lewismc — APPROVED** [review]

**Comment by lewismc** [comment]
I'm going to pull the code locally @aheermann and test. Please standby before merging thank you. Excellent job.

**Comment by lewismc** [comment]
@capstone-coal/19-usc-capstone-team please be in a position to discuss and compare results of this review by tomorrows meeting. Thank you all in advance.

**Comment by lewismc** [comment]
@aheermann please update this PR to resolve conflict. 

@capstone-coal/19-usc-capstone-team has anyone else tested this PR out?

**Comment by Luner** [comment]
During our meeting I believe @aheermann mentioned looking into the issues involving the memory leak with joblib.  He mentioned that there is a known issue within the library.

**Comment by lewismc** [comment]
@aheermann is this ready too be tested again then? Thanks for pushing most recent commit.

**Comment by lewismc** [comment]
PING here... do you have a status update please?

**Comment by aheermann** [comment]
@lewismc, this should have fixed the memory leak, as long as you make sure you have psutil installed. Past that, I imagine this branch will not be merged to master, but instead be combined with the pytorch one, for the config file branch.

**Comment by lewismc** [comment]
@aheermann I'll test this right now and provide feedback.

> Past that, I imagine this branch will not be merged to master, but instead be combined with the pytorch one, for the config file branch.

No let's merge this one first. We can then update #199 and finally #198

**Comment by lewismc** [comment]
BTW, installation on < python: stable 3.7.5  resulted in failure. Upgrading to Python 3.7.5 fixed that. Just in case anyone tries.

**Comment by lewismc** [comment]
Hi @aheermann this has fixed the issue. See below for log
```
2019-11-20 10:41:33,058 Starting mineral classification with input file 'avng.jpl.nasa.gov/AVNG_2015_data_distribution/L2/ang20150420t182050_rfl_v1e/ang20150420t182050_corr_v1e_img.hdr' and spectral library '../pycoal/tests/s07_AV95_envi.hdr'.
2019-11-20 10:41:33,058 Instantiated Mineral Classifier with following specification: -classifier function 'SAM'
2019-11-20 10:41:33,058 Starting generation of three-band RGB image from input file: 'avng.jpl.nasa.gov/AVNG_2015_data_distribution/L2/ang20150420t182050_rfl_v1e/ang20150420t182050_corr_v1e_img.hdr' with following RGB values R: '680.0', G: '532.5', B: '472.5'
2019-11-20 10:41:41,134 Saving RGB image as 'ang20150420t182050_corr_v1e_img_rgb.hdr'
2019-11-20 10:41:41,254 Completed RGB image generation. Time elapsed: '0:00:08'
2019-11-20 10:41:41,302 Starting Mineral Classification for image 'avng.jpl.nasa.gov/AVNG_2015_data_distribution/L2/ang20150420t182050_rfl_v1e/ang20150420t182050_corr_v1e_img.hdr', saving classified image to 'ang20150420t182050_corr_v1e_img_class.hdr'
2019-11-20 10:41:41,384 Classifying a 14674X739 image
2019-11-20 18:26:23,455 Completed Mineral Classification. Time elapsed: '7:44:42'
```
Excellent work here. If we take into consideration that the initial processing time was well over a day... or maybe even longer, then this represents a huge speedup. I'm going to go ahead and merge this into master branch. Excellent work.

**Pull request merged**